enhancements/helm3/console.md

```diff
@@ -0,0 +1,144 @@
+---
+title: helm-charts-in-developer-catalog
+authors:
+  - "@sbose78"
+  - "@pedjak"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-01-09
+last-updated: 2020-03-19
+status: implementable
+---
+
+# Helm Charts in the Developer Catalog
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Managing Helm charts using the Developer perspective from DevConsole
+
+## Motivation
+
+Helm is a Kubernetes package manager.  Helm 3.0 is a major release of helm which brings in a rich set of features and addresses major security concerns around tiller.  
+
+Red hat Openshift wants to bring Helm based content support to Openshift 4.4 Developer Catalog along with Operators to strengthen the helm based ecosystem.
+
+![Developer Catalog](../helm3/assets/dev-catalog.png)
+
+### Goals
+
+* Provide RESTful API for managing Helm charts and releases
+* Support disconnected installs
+* Support easy management of available charts, aggregation from multiple sources and filtering
+  
+### Non-Goals
+
+* Infrastructure for serving the default chart repository
+* Process for curating charts within the default chart repository
+  
+## Proposal
+  
+### Charts in the Developer Catalog
+
+The charts that would show up in the Developer Catalog will be powered by a [standard](https://helm.sh/docs/topics/chart_repository) helm charts' repository service.
+
+The default chart repository will be served from [Red Hat public GitHub repository](https://github.com/redhat-developer/redhat-helm-charts).
+
+New charts will be added or existing curated by submitting PRs against the above GitHub repo.
+
+### How would the UI discover the charts
+
+1. The UI would invoke `/api/helm/charts/index.yaml` endpoint to get [the list of all available charts](https://helm.sh/docs/topics/chart_repository/#the-index-file) so that the available charts can be rendered in the developer catalog. 
+
+2. The above endpoint would proxy requests to the configured charts repository
```

Comment:
a proxy surprises me.  Why wouldn't you just directly access the charts repo?

Reply:
The charts repo URL is supposed to configurable. Console UI therefore always talks to a specific console backend URL which proxies to the configured charts repo URL.

The UI doesn't have to be aware as long as the Console Backend is aware.

Note, this change is already in master and us being shipped in 4.4.

Reply:
so there's only one chart repository per cluster?  or is the proxy aggregating multiple chart repositories?

Reply:
> so there's only one chart repository per cluster? or is the proxy aggregating multiple chart repositories?

@bparees at the moment, yes, but we already have [uses cases](https://issues.redhat.com/browse/ODC-2994) that would require aggregation from multiple repos.

> and that interaction with external URL off cluster will go through configured ocp cluster egress proxy?

The enhancement does not enforce that, but it could be [added](https://docs.openshift.com/container-platform/4.3/networking/enable-cluster-wide-proxy.html) by cluster admins orthogonally, if needed, correct @derekwaynecarr ?

Reply:
> The enhancement does not enforce that, but it could be added by cluster admins orthogonally, if needed, correct @derekwaynecarr ?

Seems like the enhancement has to enforce that.  If this EP is defining a new component which has requirements to make external requests, the new component must respect the global cluster proxy configuration.

Reply:
> @bparees at the moment, yes, but we already have uses cases that would require aggregation from multiple repos.

it doesn't seem like this EP defines an api that allows for that (the api seems to be "single instance of a cluster scoped resource, and the resource instance only has a field for a single chart repository url")

Reply:
> If this EP is defining a new component which has requirements to make external requests, 

This EP defines a CR and a few endpoints that could be implemented either by existing or a new component. 

The reality is that the endpoints got implemented by console backend already.

Reply:
> it doesn't seem like this EP defines an api that allows for that (the api seems to be "single instance of a cluster scoped resource, and the resource instance only has a field for a single chart repository url")

We are proposing a new top-level CR (HelmChartRepository) that could be instantiated multiple-times. If cluster-scoped resources could be instantiated only once, we can introduce a CR that manages a list of repos (see https://github.com/openshift/enhancements/pull/175#discussion_r405446892) or move CR to `helm.openshift.io` group (as suggested/asked by @derekwaynecarr )

Reply:
> If cluster-scoped resources could be instantiated only once

they can be instantiated multiple times, but since your example named the resource `cluster` and i thought there was some discussion around only having one instance, i drew the conclusion you intended to only consume that single special instance.  If you're going to consume all instances, then i guess that's fine.